### PR TITLE
Use config values as default in `PydanticBaseEnvSettingsSource.__init__`

### PR DESCRIPTION
**Pull request by hramezani**
Fixes https://github.com/pydantic/pydantic-settings/issues/98

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by hramezani** [comment]
please review

File: tests/test_settings.py [hunk 1]
```diff
@@ -1698,3 +1699,21 @@ class Settings(BaseSettings):
...
+    class Settings(BaseSettings):
+        foo: str = 'test'
+
+        model_config = SettingsConfigDict(env_prefix='prefix_', case_sensitive=True)
+
+    s = Settings()
+    assert s.model_config['env_prefix'] == 'prefix_'
+    assert s.model_config['case_sensitive'] is True
+
+    c = CustomEnvSettingsSource(s)
```
Comment (Viicos) [inline tests/test_settings.py:1717]: `s` is an instance of `Settings`, but sources expect `type[BaseSettings]`

**Review by Viicos — COMMENTED** [review]
I don't think this addresses the issue unfortunately. Here's a MRE of what should be expected:

```python
class CustomEnvSettingsSource(EnvSettingsSource):
    pass

class Settings(BaseSettings):
    foo: str

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        custom_env_settings = CustomEnvSettingsSource(settings_cls)  # CustomEnvSettingsSource should be able to use config values either from `model_config` or values given at instantiation
        return init_settings, custom_env_settings, dotenv_settings, file_secret_settings

os.environ["prefix_foo"] = "test_foo"

s = Settings(_env_prefix="prefix_")
assert s.foo == "test_foo"
```

**Comment by hramezani** [comment]
> Thanks!
> 
> I don't think this addresses the issue unfortunately. Here's a MRE of what should be expected:
> 
> ```python
> class CustomEnvSettingsSource(EnvSettingsSource):
>     pass
> 
> class Settings(BaseSettings):
>     foo: str
> 
>     @classmethod
>     def settings_customise_sources(
>         cls,
>         settings_cls: type[BaseSettings],
>         init_settings: PydanticBaseSettingsSource,
>         env_settings: PydanticBaseSettingsSource,
>         dotenv_settings: PydanticBaseSettingsSource,
>         file_secret_settings: PydanticBaseSettingsSource
>     ) -> tuple[PydanticBaseSettingsSource, ...]:
>         custom_env_settings = CustomEnvSettingsSource(settings_cls)  # CustomEnvSettingsSource should be able to use config values either from `model_config` or values given at instantiation
>         return init_settings, custom_env_settings, dotenv_settings, file_secret_settings
> 
> os.environ["prefix_foo"] = "test_foo"
> 
> s = Settings(_env_prefix="prefix_")
> assert s.foo == "test_foo"
> ```
> 
> What the added tests show is that `CustomEnvSettingsSource` is able to read values from `model_config`, but not from instantation arguments 😕

You can pass `env_prefix` in custom source initialization.

```py
custom_env_settings = CustomEnvSettingsSource(settings_cls, env_prefix="prefix_")
```

or define it in config as I shared an example in the [issue](https://github.com/pydantic/pydantic-settings/issues/98#issuecomment-1619807220)

**Comment by Viicos** [comment]
> You can pass env_prefix in custom source initialization.
> ```python
> custom_env_settings = CustomEnvSettingsSource(settings_cls, env_prefix="prefix_")
> ```
> or define it in config as I shared an example in the https://github.com/pydantic/pydantic-settings/issues/98#issuecomment-1619807220

But the `env_prefix` is then hardcoded into the code, and not configurable from instantiation arguments

I understand the MRE I gave [here](https://github.com/pydantic/pydantic-settings/pull/100#pullrequestreview-1512353807) can't be supported, as trying to support it might lead to too many breaking changes.
I can still offer a PR and see if it's worth it 😉

**Comment by hramezani** [comment]
@Viicos `_env_prefix` and the other instantiation arguments are for settings class init not the source class init.

If you define a custom source, you need to initialize it in `settings_customise_sources` and then you can pass args there.


> I understand the MRE I gave https://github.com/pydantic/pydantic-settings/pull/100#pullrequestreview-1512353807 can't be supported, as trying to support it might lead to too many breaking changes.
I can still offer a PR and see if it's worth it 😉

I would be happy to take a look at that. we can merge it if it is an easy one. because as I mentioned above, you can initialize a custom source in `settings_customise_sources` and it is not block you.

**Review by dmontagu — APPROVED** [review]

**Review by Kludex — APPROVED** [review]

**Comment by Kludex** [comment]
I merge without seeing the messages above, sorry. Hope is cool.

**Comment by Viicos** [comment]
> I merged without seeing the messages above, sorry. Hope is cool.

Yes this is still an improvement, thanks for merging

**Pull request merged**